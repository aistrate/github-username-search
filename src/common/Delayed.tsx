import { useEffect, useState } from "react";

export default Delayed;

type DelayedProps = {
  delay?: number;
  children?: React.ReactNode;
};

function Delayed({ delay = 500, children }: DelayedProps) {
  const [isVisible, setIsVisible] = useState(false);

  useEffect(() => {
    const timeout = window.setTimeout(() => {
      setIsVisible(true);
    }, delay);

    return () => {
      clearTimeout(timeout);
    };
  }, [delay]);

  return <>{isVisible && <>{children}</>}</>;
}
