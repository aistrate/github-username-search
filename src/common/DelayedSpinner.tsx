import { useEffect, useState } from "react";
import Spinner from "./styled/Spinner";

export default DelayedSpinner;

type DelayedSpinnerProps = {
  delay?: number;
};

function DelayedSpinner({ delay = 500 }: DelayedSpinnerProps) {
  const [isVisible, setIsVisible] = useState(false);

  useEffect(() => {
    const timeout = window.setTimeout(() => {
      setIsVisible(true);
    }, delay);

    return () => {
      clearTimeout(timeout);
    };
  }, [delay]);

  return <>{isVisible && <Spinner />}</>;
}
