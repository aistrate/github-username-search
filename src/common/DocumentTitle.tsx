import { useEffect } from "react";

export default DocumentTitle;

type DocumentTitleProps = {
  value: string;
};

function DocumentTitle({ value }: DocumentTitleProps) {
  useEffect(() => {
    document.title = `${value} - ${process.env.REACT_APP_APPLICATION_NAME}`;
  }, [value]);

  return null;
}
