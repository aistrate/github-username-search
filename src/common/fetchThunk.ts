import type { ActionReducerMapBuilder, AsyncThunk } from "@reduxjs/toolkit";
import { createAsyncThunk } from "@reduxjs/toolkit";

type FetchState<Data> = {
  data?: Data;
  error?: string;
  httpStatus?: number;
  requestUrl?: string;
  isLoading: boolean;
};

const createFetchThunk = <Data, ThunkArg, ResponseData>(
  typePrefix: string,
  getRequestUrl: (arg: ThunkArg) => string | null,
  extractData: (responseData: ResponseData) => Data
) =>
  createAsyncThunk<
    FetchState<Data>,
    ThunkArg,
    { rejectValue: FetchState<Data> }
  >(typePrefix, async (thunkArg, { rejectWithValue }) => {
    const requestUrl = getRequestUrl(thunkArg);

    if (!requestUrl) {
      return { isLoading: false };
    }

    let response: Response;
    try {
      response = await fetch(requestUrl, fetchOptions);
    } catch (err) {
      const error = `Error: ${(err as Error).message}`;
      return rejectWithValue({ error, requestUrl, isLoading: false });
    }

    if (!response.ok) {
      const errorData = await response.json();
      const error = `HTTP Error: (${response.status}) ${errorData.message}`;
      return rejectWithValue({
        error,
        httpStatus: response.status,
        requestUrl,
        isLoading: false,
      });
    }

    const responseData: ResponseData = await response.json();
    return {
      data: extractData(responseData),
      httpStatus: response.status,
      requestUrl,
      isLoading: false,
    };
  });

// define the environment variable in file .env.development.local (NOT production);
// the file should not be tracked by source control (add to .gitignore)
const auth = process.env.REACT_APP_GITHUB_API_AUTH;

// this will increase the GitHub API rate limit from 60 to 5000 requests/hour
const fetchOptions = auth
  ? { headers: new Headers({ Authorization: auth }) }
  : undefined;

function addFetchCaseReducers<Data, ThunkArg>(
  builder: ActionReducerMapBuilder<FetchState<Data>>,
  fetchThunk: AsyncThunk<
    FetchState<Data>,
    ThunkArg,
    { rejectValue: FetchState<Data> }
  >
) {
  builder
    .addCase(fetchThunk.pending, (state) => {
      return {
        data: state.data, // to avoid flickering
        isLoading: true,
      };
    })
    .addCase(fetchThunk.fulfilled, (_state, action) => {
      return action.payload;
    })
    .addCase(fetchThunk.rejected, (_state, action) => {
      return action.payload;
    });
}

export type { FetchState };

export { createFetchThunk, addFetchCaseReducers };
