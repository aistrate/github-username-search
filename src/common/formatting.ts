export { formatDateTime, formatNumber, formatUrl, setLocale };

let locale = "en-GB";

function setLocale(value: string) {
  locale = value;
}

const dateOptions: Intl.DateTimeFormatOptions = {
  year: "numeric",
  month: "short",
  day: "numeric",
};

const dateTimeOptions: Intl.DateTimeFormatOptions = {
  ...dateOptions,
  hour: "2-digit",
  minute: "2-digit",
};

const weekdayDateTimeOptions: Intl.DateTimeFormatOptions = {
  ...dateTimeOptions,
  weekday: "short",
};

type DateTimeFormat = "date" | "date-time" | "weekday-date-time";

function formatDateTime(value: string | number, format: DateTimeFormat) {
  return new Date(value).toLocaleDateString(locale, getDateTimeOptions(format));
}

function getDateTimeOptions(format: DateTimeFormat) {
  switch (format) {
    case "date":
      return dateOptions;

    case "date-time":
      return dateTimeOptions;

    case "weekday-date-time":
      return weekdayDateTimeOptions;

    default:
      const _never: never = format;
      return _never;
  }
}

function formatNumber(value: number) {
  return value.toLocaleString(locale);
}

const urlWithProtocol = /^https?:\/\//i;

function formatUrl(url: string) {
  return url.match(urlWithProtocol) ? url : `http://${url}`;
}
