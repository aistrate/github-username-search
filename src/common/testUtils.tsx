import { render } from "@testing-library/react";
import { ReactElement, ReactNode } from "react";
import { Provider } from "react-redux";
import { MemoryRouter, useLocation } from "react-router-dom";
import { ReactTestRendererJSON } from "react-test-renderer";
import styled from "styled-components";
import { createStore } from "../app/store";

export { renderWithWrapper, RoutingLocation, removeClassNames };

function renderWithWrapper(ui: ReactElement, initialRoute = "/") {
  const store = createStore();

  const Wrapper = ({ children }: { children: ReactNode }) => {
    return (
      <Provider store={store}>
        <MemoryRouter initialEntries={[initialRoute]}>{children}</MemoryRouter>
      </Provider>
    );
  };

  return render(ui, { wrapper: Wrapper });
}

function RoutingLocation() {
  const location = useLocation();

  return (
    <Container data-testid="routingLocation">
      {location.pathname + location.search}
    </Container>
  );
}

const Container = styled.div`
  position: fixed;
`;

/**
 * CSS class names generated by styled-components are non-deterministic,
 * so they will interfere with snapshot testing.
 */
function removeClassNames(
  tree: null | ReactTestRendererJSON | ReactTestRendererJSON[]
) {
  if (!tree) return;

  if (!Array.isArray(tree)) {
    tree = [tree];
  }

  tree.forEach((n) => {
    if (typeof n === "object" && n.type) {
      if (n.props && n.props.className) {
        delete n.props.className;
      }

      if (n.children) {
        n.children.forEach((child) => {
          if (typeof child !== "string") {
            removeClassNames(child);
          }
        });
      }
    }
  });
}
