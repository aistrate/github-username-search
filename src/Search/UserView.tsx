import React from "react";
import styled, { css } from "styled-components/macro";
import type { UserType, User } from "./Models";
import type { FetchResult } from "../Shared/Fetch";
import { formatDateTime, formatNumber } from "../Shared/Formatting";
import { DelayedSpinner } from "../Shared/DelayedSpinner";
import { Message } from "../Styled/Message";
import { LargeHeading } from "../Styled/Headings";
import { Row } from "../Styled/Row";
import { ExternalLink } from "../Styled/Links";

export default UserView;

type UserViewProps = {
  userFetch: FetchResult<User>;
  username: string;
  className?: string;
};

function UserView({ userFetch, username, className }: UserViewProps) {
  return (
    <Container className={className}>
      {userFetch.isLoading && <DelayedSpinner />}

      {userFetch.error &&
        (userFetch.httpStatus === 404 ? (
          <Message type="info">{`Username '${username}' was not found.`}</Message>
        ) : (
          <Message type="error">{userFetch.error}</Message>
        ))}

      {userFetch.data && <UserInfo user={userFetch.data} />}
    </Container>
  );
}

type UserInfoProps = {
  user: User;
};

function UserInfo({ user }: UserInfoProps) {
  return (
    <>
      <Avatar url={user.avatar_url} userType={user.type} />

      <LargeHeading>{user.name || user.login}</LargeHeading>

      <dl>
        <Row label="Username">
          <ExternalLink href={user.html_url}>{user.login}</ExternalLink>
        </Row>
        <Row label="Type">{user.type}</Row>
        <Row label="Created">{formatDateTime(user.created_at, "date")}</Row>
        {user.bio && <Row label="Bio">{user.bio}</Row>}
        {user.followers > 0 && (
          <Row label="Followers">{formatNumber(user.followers)}</Row>
        )}
        {user.company && (
          <Row label="Company">
            <CompanyList names={user.company} />
          </Row>
        )}
        {user.location && <Row label="Location">{user.location}</Row>}
        {user.email && <Row label="Email">{user.email}</Row>}
        {user.blog && (
          <Row label="Blog">
            <ExternalLink nowrap href={user.blog}>
              {user.blog}
            </ExternalLink>
          </Row>
        )}
        {user.twitter_username && (
          <Row label="Twitter">
            <ExternalLink href={`https://twitter.com/${user.twitter_username}`}>
              @{user.twitter_username}
            </ExternalLink>
          </Row>
        )}
        <Row label="Repositories">{formatNumber(user.public_repos)}</Row>
      </dl>
    </>
  );
}

type AvatarProps = {
  url: string;
  userType: UserType;
};

function Avatar({ url, userType }: AvatarProps) {
  const size = userType === "User" ? 180 : 100;

  return (
    <Image
      key={url}
      src={url}
      alt="Avatar"
      width={size}
      height={size}
      circular={userType === "User"}
    />
  );
}

type CompanyListProps = {
  names: string;
};

function CompanyList({ names }: CompanyListProps) {
  // 'names' can be a comma-separated list
  const companyList = names
    .split(",")
    .map((name) => name.trim())
    .map((name) => <Company name={name} />);

  const joined = companyList
    .slice(1)
    .reduce((accumulator, company) => accumulator.concat(<>, </>, company), [
      companyList[0],
    ]);

  // use createElement with spread argument to avoid React's warning:
  // 'Each child in a list should have a unique "key" prop'
  return React.createElement(React.Fragment, null, ...joined);
}

type CompanyProps = {
  name: string;
};

function Company({ name }: CompanyProps) {
  return name[0] === "@" ? (
    <ExternalLink href={`https://github.com/${name.slice(1)}`}>
      {name}
    </ExternalLink>
  ) : (
    <>{name}</>
  );
}

const Container = styled.div`
  min-height: 60px;
  position: relative;
`;

type ImageProps = {
  circular: boolean;
};

const Image = styled.img<ImageProps>`
  height: auto;
  vertical-align: middle;

  ${({ circular }) =>
    circular &&
    css`
      border-radius: 50%;
    `}
`;
