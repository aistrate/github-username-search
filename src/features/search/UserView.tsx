import React from "react";
import styled, { css } from "styled-components/macro";
import Delayed from "../../common/Delayed";
import { formatDateTime, formatNumber } from "../../common/formatting";
import { LargeHeading } from "../../common/styled/Headings";
import { ExternalLink } from "../../common/styled/Links";
import Message from "../../common/styled/Message";
import Row from "../../common/styled/Row";
import Spinner from "../../common/styled/Spinner";
import { User, UserFetch, UserType } from "./userSlice";

export default UserView;

type UserViewProps = {
  userFetch: UserFetch;
  className?: string;
};

function UserView({ userFetch, className }: UserViewProps) {
  const username = userFetch.fetchArg?.username || "";

  return (
    <Container className={className}>
      {userFetch.isLoading && (
        <Delayed>
          <Spinner />
        </Delayed>
      )}

      {userFetch.error &&
        (userFetch.httpStatus === 404 ? (
          <Message type="info">{`Username '${username}' was not found.`}</Message>
        ) : (
          <Message type="error">{userFetch.error}</Message>
        ))}

      {userFetch.data && <UserInfo user={userFetch.data} />}
    </Container>
  );
}

type UserInfoProps = {
  user: User;
};

const UserInfo = React.memo(({ user }: UserInfoProps) => {
  return (
    <>
      <Avatar url={user.avatar_url} userType={user.type} />

      <LargeHeading>{user.name || user.login}</LargeHeading>

      <dl>
        <Row label="Username">
          <ExternalLink href={user.html_url}>{user.login}</ExternalLink>
        </Row>
        <Row label="Type">{user.type}</Row>
        <Row label="Created">{formatDateTime(user.created_at, "date")}</Row>
        {user.bio && <Row label="Bio">{user.bio}</Row>}
        {user.followers > 0 && (
          <Row label="Followers">{formatNumber(user.followers)}</Row>
        )}
        {user.company && (
          <Row label="Company">
            <CompanyList names={user.company} />
          </Row>
        )}
        {user.location && <Row label="Location">{user.location}</Row>}
        {user.email && <Row label="Email">{user.email}</Row>}
        {user.blog && (
          <Row label="Blog">
            <ExternalLink nowrap href={user.blog}>
              {user.blog}
            </ExternalLink>
          </Row>
        )}
        {user.twitter_username && (
          <Row label="Twitter">
            <ExternalLink href={`https://twitter.com/${user.twitter_username}`}>
              @{user.twitter_username}
            </ExternalLink>
          </Row>
        )}
        <Row label="Repositories">{formatNumber(user.public_repos)}</Row>
      </dl>
    </>
  );
});

type AvatarProps = {
  url: string;
  userType: UserType;
};

function Avatar({ url, userType }: AvatarProps) {
  const size = userType === "User" ? 180 : 100;

  return (
    <Image
      key={url}
      src={url}
      alt="Avatar"
      width={size}
      height={size}
      circular={userType === "User"}
    />
  );
}

type CompanyListProps = {
  names: string;
};

function CompanyList({ names }: CompanyListProps) {
  // 'names' can be a comma-separated list
  const companyList = names
    .split(",")
    .map((name) => name.trim())
    .map((name) => <Company name={name} />);

  if (companyList.length === 0) return <></>;

  const comma = <>{", "}</>;

  const joined = companyList
    .slice(1)
    .reduce(
      (accumulator, company) => accumulator.concat(comma, company),
      [companyList[0]]
    );

  // use createElement with spread argument to avoid React's warning:
  // 'Each child in a list should have a unique "key" prop'
  return React.createElement(React.Fragment, null, ...joined);
}

type CompanyProps = {
  name: string;
};

function Company({ name }: CompanyProps) {
  return name[0] === "@" ? (
    <ExternalLink href={`https://github.com/${name.slice(1)}`}>
      {name}
    </ExternalLink>
  ) : (
    <>{name}</>
  );
}

const Container = styled.div`
  min-height: 60px;
  position: relative;
`;

type ImageProps = {
  circular: boolean;
};

const Image = styled.img<ImageProps>`
  display: block;
  height: auto;

  ${({ circular }) =>
    circular &&
    css`
      border-radius: 50%;
    `}
`;
