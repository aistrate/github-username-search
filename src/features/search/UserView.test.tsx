import renderer from "react-test-renderer";
import { removeClassNames } from "../../common/testUtils";
import { CompanyList } from "./UserView";

test("<CompanyList> transforms company names prefixed with '@' into links to each company's GitHub username", () => {
  let tree = renderer
    .create(<CompanyList names="@Google, @Microsoft, OtherCompany" />)
    .toJSON();

  removeClassNames(tree);

  expect(tree).toMatchSnapshot();
});
