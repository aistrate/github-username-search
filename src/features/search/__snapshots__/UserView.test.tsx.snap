// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<CompanyList> transforms company names prefixed with '@' into links to each company's GitHub username 1`] = `
Array [
  <a
    href="https://github.com/Google"
    rel="noreferrer"
    target="_blank"
  >
    @Google
  </a>,
  ", ",
  <a
    href="https://github.com/Microsoft"
    rel="noreferrer"
    target="_blank"
  >
    @Microsoft
  </a>,
  ", ",
  "OtherCompany",
]
`;
